### PR TITLE
Import kerberos-login-module

### PR DESCRIPTION
**Pull request by zpqrtbnk**
This PR imports sample code for a Kerberos login module, corresponding to Kerberos support in the .NET Client 4.0. This is a DRAFT PR, while it is being reviewed.

Notes to self: in order to build, use the Maven window in IntelliJ to trigger `compile` and then `package`.

### CODE REVIEW

File: enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    public boolean onLogin() throws LoginException {
+
+        // get the token
+        byte[] token = getToken();
+
+        // accept the token
+        GSSContext gssContext = acceptToken(token);
+
+        // validate access (throws if access is denied)
+        validateAccess(gssContext);
```
Comment (kwart) [inline enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java]: I would rename the method to a more descriptive name here and wrap it into the `skipRole` condition.
E.g.

```java
if (! isSkipRole()) {
  addGroupSidsFromMsPac(token);
}
```

File: enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+                pacAuthzEntry = null;
+            }
+
+            if (authzEntry == null) {
+                throw new FailedLoginException("Internal error (no AD_IF_RELEVANT entry).");
+            }
+
+            if (pacAuthzEntry == null) {
+                throw new FailedLoginException("Internal error (no AD-WIN2K-PAC entry within the AD-IF-RELEVANT entry).");
+            }
```
Comment (kwart) [inline enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java]: I think we don't need these checks. The missing role means no permission mapped, so we should be safe. If you still want to have the checks included, I would suggest renaming the project from `kerberos-login-module` to `active-directory-login-module`.

File: enterprise/kerberos-login-module/pom.xml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+          <artifactId>kerberos-login-module</artifactId>
+          <version>${hazelcast.version}</version>
+          <configuration>
+            <archive>
+              <manifestFile>src/main/resources/META-INF/MANIFEST.MF</manifestFile>
+            </archive>
+          </configuration>
+        </plugin>
+      </plugins>
+    </build>
```
Comment (kwart) [inline enterprise/kerberos-login-module/pom.xml]: Probably we don't need this part. Od do you implement some Maven plugin within the code?

File: enterprise/kerberos-login-module/pom.xml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            <version>${project.version}</version>
+        </dependency>
+        
+        <dependency>
+          <groupId>KerberosUtils</groupId>
+          <artifactId>KerberosUtils</artifactId>
+          <version>0.0</version>
+          <scope>system</scope>
+          <systemPath>${project.basedir}/lib/KerberosUtils.jar</systemPath>
+        </dependency>
```
Comment (kwart) [inline enterprise/kerberos-login-module/pom.xml]: We should avoid system dependencies.

File: enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+package com.hazelcast.security;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.auth.login.FailedLoginException;
+import javax.security.auth.login.LoginException;
+
+import com.hazelcast.internal.util.StringUtil;
+import com.sun.security.jgss.ExtendedGSSContext;
+import com.sun.security.jgss.InquireType;
```
Comment (kwart) [inline enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java]: The README should mention the LoginModule requires Oracle/OpenJDK JVM.

File: enterprise/kerberos-login-module/pom.xml (hunk 3 above)
Comment (zpqrtbnk) [reply enterprise/kerberos-login-module/pom.xml]: Realized we don't need it indeed, removing.

File: enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java (hunk 5 above)
Comment (zpqrtbnk) [reply enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java]: Adding to README.

File: enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java (hunk 2 above)
Comment (zpqrtbnk) [reply enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java]: Makes sense. Fixing: if we cannot find the entries, we just don't map any more.

File: enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java (hunk 1 above)
Comment (zpqrtbnk) [reply enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java]: Understood. Refactoring.

File: enterprise/kerberos-login-module/pom.xml (hunk 4 above)
Comment (zpqrtbnk) [reply enterprise/kerberos-login-module/pom.xml]: Understood. I am using classe for [this library](https://github.com/jcmturner/java-kerberos-utils) which is published on GitHub but does not seem to be available on a Maven repository of some sort - what would be the best solution here? Copy & embed the code in our module? Or?

Comment (zpqrtbnk) [reply enterprise/kerberos-login-module/pom.xml]: Have experimented with JaasLounge and other libs - all libs are quite old, which is ... weird ... anyways the simplest solution for now seems to be to embed the bare minimum of the KerberosUtils lib I was using. So, have updated the PR.

File: enterprise/kerberos-login-module/README.md [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+http://hazelcast.com/hazelcast-enterprise-trial/
+
+## Introduction
+
+Demonstrates how to create a Kerberos Login Module for Hazelcast.
+
+The project builds and produces the `kerberos-login-module.jar` artifact. It relies on the [Java Kerberos Utilities library](https://github.com/jcmturner/java-kerberos-utils) which is available under the Apache License 2.0. The library is included in the project's `lib` directory, and embedded in the produced artifact.
+
+The module relies on Java GSS and therefore requires Oracle/OpenJDK JVM.
```
Comment (kwart) [inline enterprise/kerberos-login-module/README.md]: The JGSS is available in IBM JDKs too, the problem is in using private `com.sun.security` and `sun.security` packages. So I would choose another wording here. E.g.

> The module relies on internal Java GSS implementation classes in `com.sun.security` and `sun.security` packages and therefore requires Oracle/OpenJDK JVM.

Comment (zpqrtbnk) [reply enterprise/kerberos-login-module/README.md]: Fixing

**Comment by kwart** [comment]
The build fails on checkstyle errors. Probably it will need some exclusion rules due to the imported source files.

**Comment by mmedenjak** [comment]
@zpqrtbnk @kwart can we merge this PR or progress on it somehow?

**Comment by zpqrtbnk** [comment]
This was an experimental PR while we developed Kerberos support on top of IMDG 4.0, but now a different and more stable / official solution is available as part of IMDG 4.1 which officially supports Kerberos, so this PR is now moot. Closing.

**Pull request closed without merging**